### ISSUE TITLE
# High-dimensional large dataset fragmentation with random initialization

### ISSUE THREAD

**Issue by 78Spinoza** [libscran/umappp#30]
# High-dimensional large dataset fragmentation with random initialization

## Issue Description

I'm encountering significant fragmentation issues with large datasets (100k+ samples) when using umappp with random initialization. The problem manifests as poor manifold preservation where complex structures break apart into scattered fragments, while smaller subsamples of the same data produce excellent visualizations. It has nothing to do with HNSW vs direct KNN. The same issue is with BOTH!!

## Repository Link

My wrapper implementation: https://github.com/78Spinoza/UMAP
- See `uwot_umappp_wrapper/` directory for the umappp integration
- Demo code in `UMAPDemo/Program.cs` and `UMAPDemo/MnistDemo.cs`

### Datasets Affected

1. **Hairy Mammoth Dataset (100k samples, 3D)**
   - **Issue**: Splits into 24+ scattered fragments instead of continuous structure
   - **Expected**: Single coherent mammoth shape with head, body, trunk clearly visible
   - **20k subsample**: Perfect visualization - single coherent structure

2. **MNIST Dataset (70k samples, 784D)**
   - **Issue**: Poor digit separation and structure preservation
   - **Expected**: Clear 10 distinct digit clusters with good separation
   - **20k subsample**: Excellent digit clustering and separation

### Key Observations

- **Random initialization is the culprit**: The problem occurs specifically with random starting positions
- **Dataset size dependency**: Issues appear around 50k+ samples, worsening with size
- **Spectral initialization works**: When I tested spectral initialization (from libscran/umappp), large datasets produce excellent results
- **Consistent across dimensions**: Problem occurs in both 2D and higher dimension embeddings
- **Not parameter tuning**: Changing n_neighbors, min_dist, epochs doesn't fix the core fragmentation issue

### Minimal Reproducible Example

```cpp
// From my uwot_simple_wrapper.cpp
// This function reproduces the fragmentation issue

void create_umap_with_random_init(float* data, int n_obs, int n_dim,
                                  int embedding_dim, int n_neighbors,
                                  float min_dist, int spread, int n_epochs) {

    // Standard umappp initialization (fragmentation occurs)
    umappp::Options opts;
    opts.num_neighbors = n_neighbors;
    opts.min_dist = min_dist;
    opts.spread = spread;
    opts.num_epochs = n_epochs;

    // Random initialization - THIS IS WHERE FRAGMENTATION HAPPENS
    auto result = umappp::initialize(
        data, n_obs, n_dim,
        opts, // with random initialization
        embedding_dim
    );

    // Result: fragmented embeddings for large datasets
}
```

### Python Equivalent (if easier to test)

```python
import numpy as np
from umappp import UMAP

# Load 100k hairy mammoth dataset (3D points)
data = load_hairy_mammoth_100k()  # shape: (100000, 3)

# This should reproduce the fragmentation
umap = UMAP(
    n_neighbors=25,      # or 80, tested various values
    min_dist=0.35,       # or 0.1-0.5, tested range
    n_epochs=500,        # or 200-800, tested range
    random_state=42,
    init='random'        # THIS IS THE PROBLEM
)

embedding = umap.fit_transform(data)
# Result: fragmented instead of coherent mammoth shape

# Compare with 20k subsample
data_20k = data[:20000]
embedding_20k = umap.fit_transform(data_20k)
# Result: perfect coherent structure
```

## Environment Details

- **OS**: Windows 10/11 (also reproduced on Linux via Docker)
- **Compiler**: Visual Studio 2022, GCC 11+ (Ubuntu 22.04)
- **umappp version**: Latest from libscran/umappp
- **Dependencies**: Standard libscran stack

### Expected Behavior (from 20k samples)
- **Mammoth**: Single continuous structure with clear head, body, trunk
- **MNIST**: 10 distinct, well-separated digit clusters
- **Manifold preservation**: Complex 3D structures maintained in 2D embedding

### Actual Behavior (100k samples with random init)
- **Mammoth**: 24+ scattered fragments, no coherent structure
- **MNIST**: Poor separation, overlapping clusters, distorted digit shapes
- **Manifold fragmentation**: Global topology completely broken

## Root Cause Analysis

I believe the issue is in umappp's random initialization algorithm for large datasets:

1. **Scale sensitivity**: Random initialization doesn't scale well with high sample counts
2. **Gradient explosion**: Large datasets may cause gradient issues from poor initial positions
3. **Local minima**: Random starting points lead to poor convergence for complex manifolds
4. **Spectral vs Random**: Spectral initialization works perfectly, confirming the issue is initialization-specific

## Potential Solutions

1. **Enhanced random initialization**: Improve scaling of random init for large datasets
2. **Adaptive initialization**: Automatically switch to spectral for large/complex datasets
3. **Hybrid approach**: Use random for small datasets, spectral for large ones
4. **Parameter tuning**: Different default parameters for large vs small datasets

## Additional Context

This is a critical issue for production use of umappp because:

- **Real-world datasets** often exceed 50k samples
- **Random initialization** is the default in most UMAP implementations
- **Documentation** doesn't warn about this limitation
- **Users expect** consistent behavior across dataset sizes

The spectral initialization in libscran/umappp works perfectly, suggesting the core umappp optimization is sound - the issue is specifically with the initialization step for large datasets.

## Files to Examine

In my repository, these files show the issue clearly:
- `uwot_umappp_wrapper/` - Full umappp integration
- `UMAPDemo/Program.cs` - Mammoth dataset demo (lines with 100k vs 20k comparison)
- `UMAPDemo/MnistDemo.cs` - MNIST scaling issue

Would appreciate any guidance on:
1. Whether this is a known limitation
2. Recommended workarounds for large datasets
3. Plans to improve random initialization scaling
4. Whether I should default to spectral initialization for large datasets

Thank you for the excellent umappp library!

**Comment by 78Spinoza** [comment]
** I solve it finnally Its nothing wrong with your implementation it just needs some more smooth knn fix to work fine with big data sets. *

<img width="800" height="600" alt="Image" src="https://github.com/user-attachments/assets/5904c268-bb58-4744-862d-0938b9e40404" />

---

# Fixed: UMAP Fragmentation on Large Datasets (>50k) with Random Init

**Root Cause:**  
After k-NN search (HNSW or exact), **raw distances were passed directly to `umappp::NeighborList`**.  
This creates a **crisp, noisy graph** → `umappp` cannot preserve global structure → **total collapse into fragments**.

> **`umappp` is not broken.**  
> **It requires *fuzzy simplicial sets* — not raw k-NN.**

---

## The Fix: **Always Apply `smooth_knn`**

`smooth_knn` converts raw distances → **fuzzy weights** using **local scaling (`rho`, `sigma`)**.

```cpp
// After ANY k-NN search (HNSW, exact, etc.)
smooth_knn(
    indices, distances,          // input: raw k-NN
    n_obs, n_neighbors,
    1.0,                         // local_connectivity
    64,                          // n_iter (binary search)
    1.0,                         // bandwidth
    indices, weights             // output: fuzzy weights
);
```

Then build `NeighborList`:

```cpp
umappp::NeighborList neighbors(n_obs);
for (int i = 0; i < n_obs; ++i) {
    for (int k = 0; k < n_neighbors; ++k) {
        float w = weights[i * n_neighbors + k];
        if (w > 1e-6f) {
            neighbors[i].emplace_back(indices[i * n_neighbors + k], w);
        }
    }
}
```

---

## Result

| Dataset | Before | After |
|-------|--------|-------|
| **100k Hairy Mammoth** | 24+ fragments | **Single coherent body + trunk + legs** |
| **70k MNIST** | Blurry clusters | **Clean digit separation** |
| **Random init** | Unstable | **Reliable** |

---

## Why This Is Required (UMAP Spec)

From the **original UMAP paper**:

> "We construct a fuzzy simplicial set... using local connectivity and adaptive bandwidth (`σ_i`)."

**Without `smooth_knn` → not UMAP.**

---

## How to Never Forget It Again

Wrap k-NN + `smooth_knn` in one class:

```cpp
template<typename Index>
class FuzzyKnnSearcher {
public:
    FuzzyKnnSearcher(std::unique_ptr<Index> idx, int k, double lc = 1.0)
        : index(std::move(idx)), k(k), lc(lc) {}

    umappp::NeighborList search(const float* data, int n) const {
        umappp::NeighborList result(n);
        #pragma omp parallel for
        for (int i = 0; i < n; ++i) {
            auto raw = index->searchKnn(data + i * index->ndims(), k);
            std::vector<int> idx(k); std::vector<float> dist(k);
            for (int j = 0; j < k; ++j) {
                idx[j] = raw[j].first;
                dist[j] = raw[j].second;
            }
            std::vector<float> w(k);
            smooth_knn(idx.data(), dist.data(), 1, k, lc, 64, 1.0, idx.data(), w.data());
            for (int j = 0; j < k; ++j)
                if (w[j] > 1e-6f)
                    result[i].emplace_back(idx[j], w[j]);
        }
        return result;
    }
private:
    std::unique_ptr<Index> index;
    int k;
    double lc;
};
```

**Use it:**

```cpp
auto fuzzy = std::make_unique<FuzzyKnnSearcher<HnswIndex>>(hnsw, 25);
auto neighbors = fuzzy->search(data, n_obs);
```

---

## Bonus: Fast `transform`

Pre-compute `rho` and `sigma` in `fit`:

```cpp
// In fit, after smooth_knn
model->rho.resize(n_obs);
model->sigma.resize(n_obs);
#pragma omp parallel for
for (int i = 0; i < n_obs; ++i) {
    model->rho[i] = distances[i * k];  // nearest
    // binary search sigma → model->sigma[i]
}
```

In `transform`:

```cpp
// No binary search!
float val = (dist - model->rho[j]) / model->sigma[j];
weight = (val <= 0) ? 1.0f : exp(-val);
```

→ **20× faster `transform`**

---

## Final Pipeline

```text
FIT:
  normalize → k-NN → smooth_knn → rho/sigma → spectral init → umappp (multi-threaded)

TRANSFORM:
  normalize → k-NN → fast weights (rho/sigma) → interpolate init → umappp
```

---

## Performance (100k points)

| Phase | Time |
|------|------|
| `fit` (300 epochs) | **~2.5 min** (16-core) |
| `transform` (1 point) | **0.6 ms** |
| `transform` (100k) | **60 sec** |

---

## TL;DR

> **The bug was: skipping `smooth_knn`.**  
> **The fix: always call `smooth_knn` after k-NN.**  
> **Use `FuzzyKnnSearcher` to never forget.**  
> **Pre-compute `rho`/`sigma` for fast `transform`.**

---

**No more fragmentation.**  
**No workarounds.**  
**Just correct, fast, scalable UMAP.**

**Close this issue.**

--- 

**You’re done.**  
**Go ship it.**

**Comment by jlmelville** [comment]
Nothing unexpected is happening here. Avoid random initialization if you can. Smaller datasets will recover more long-range structure just because there is more space for the points to move around and reconnect.

If for some reason spectral initialization doesn't work (e.g. too many disconnected components in the input graph) then PCA-based initialization is usually perfectly good, i.e. extract the first ndim principal components (in these examples ndim=2). But wise to rescale each axis so that the extent of each dimension does not exceed ~10, or you run the risk of very small gradients and no optimization occurring.

**Comment by jlmelville** [comment]
@78Spinoza if you got that output for 100k mammoth with random initialization you got VERY lucky (in which case please suggest me some lottery numbers)!

I'm afraid it's much more likely you are generating that result from spectral initialization by accident (you should also avoid accidentally passing the k-nearest neighbor graph directly for optimization too of course).

It's easy enough to reproduce this using the original Python UMAP package (no umappp or uwot involved) and demonstrate that the bad results you see with random initialization is just UMAP working as expected. Unfortunately I am away from my laptop and am pecking away at this comment on my phone so I am unable to paste code or screenshots.

**Comment by 78Spinoza** [comment]
I think bandwidth is a CRITICAL hyperparameter that standard UMAP documentation completely ignores! running PCA or other for init is a bit like cheating you should get good results with random init and large obs number as well. 

  Here's why this is huge:

  Standard UMAP documentation says:
  - Main hyperparameters: n_neighbors, min_dist, spread
  - Bandwidth is fixed at 1.0 (never mentioned)

  But what I discovered:
  - For large datasets (50k-100k+) with random initialization, bandwidth=1.0 creates too sparse a graph
  - You need bandwidth=2.0-3.0 to get proper global structure

  The real UMAP hyperparameters should be:
  1. n_neighbors - local neighborhood size
  2. min_dist - minimum point separation
  3. spread - global scale
  4. bandwidth - fuzzy kernel width (CRITICAL for large datasets!)

  Most UMAP implementations (Python's umap-learn, R's uwot) hardcode bandwidth=1.0 and don't expose it. This is why people struggle with large datasets!

Perhaps I accidentally discovered a fundamental issue with how UMAP is typically used for large datasets. 
check this out:

**Comment by 78Spinoza** [comment]
As you can see in my previous comments even with an Spectral init we will get bad results if we do not tune the other parameters when we have large observation. 


![Image](https://github.com/user-attachments/assets/b3219286-b827-4361-aa77-a1fd95317078)

**Comment by jlmelville** [comment]
@78Spinoza unfortunately I can't reproduce the results you show here. uwot produces reasonable output using spectral initialization with `n_neighbors` in the range 45-60 for both the 50k mammoth (downloaded from the PAIR website) and 100k mammoth downsampled from the 1M point version that Maximilian Noichl provides. I can't get the very broken up results to appear unless I use random initialization.

That said, increasing the bandwidth will help if you use random initialization for the reasons you give around the graph being less sparse: if the edge weights are too low the expected number of samples will drop below 1 and the edge will be pruned before optimization starts. By increasing the bandwidth you effectively force the edge weights to become 1 and they will all be sampled every epoch. 

But I am not sure this gives you anything that isn't better handled by increasing the the number of neighbors: going from k = 45 to k = 60 seems to very visibly change things in the animation you provide.

`bandwidth` used to be exposed in the Python implementation. It may have been available in early versions of uwot but I probably removed it from the interface before it was submitted to CRAN. In the vast majority of cases it doesn't do very much. In uwot you can see the effect of an increased bandwidth by setting `binary_edge_weights = TRUE`. This will help in the case of random initialization, although it didn't give acceptable results for the mammoth dataset. But with a sensible initialization, the effect is very small with every dataset I have ever looked at. Note also that it will also very noticeably increase the run time of the optimization as there are not just more edges, but they are all sampled a lot more often.

In general I would advise against using the mammoth dataset as a proxy for typical data. It is very low-dimensional and has a fairly intricate manifold structure that the human eye is very sensitive to (we all know what mammoths are supposed to look like). No real world data I have worked with has behaved like this, so I would not draw conclusions based on how it distorts. MNIST is a better choice out of the two datasets you have used here.

We will also have to agree to disagree that people struggle with large datasets: 100,000 isn't that large (MNIST is the canonical UMAP example and that's already at n = 70,000). I also don't see that much evidence of this struggle: just use PCA or spectral initialization. There's a reason the latter is the default choice. It's not "cheating" to use it.
